controller/array_action/array_mediator_ds8k.py

```diff
@@ -0,0 +1,411 @@
+from hashlib import sha1
+import math
+from packaging.version import parse
+from pyds8k.exceptions import ClientError, ServerError, ClientException, NotFound
+from controller.common.csi_logger import get_stdout_logger
+from controller.common.size_converter import convert_size_bytes_to_gib
+from controller.common.size_converter import convert_size_gib_to_bytes
+from controller.array_action.array_mediator_interface import ArrayMediator
+from controller.array_action.utils import classproperty
+from controller.array_action.ds8k_rest_client import RESTClient
+import controller.array_action.errors as controller_errors
+from controller.array_action import config
+from controller.array_action.array_action_types import Volume
+
+
+logger = get_stdout_logger()
+
+
+# system
+SYSTEM_ID = 'id'   # 2107-{sn}
```

Comment:
But you did copy/paste from 3-rd party code. This is not right.

Reply:
But they are consts. And I'm also the writer of pyds8k, so it ’s not surprising that these two styles are similar.

Reply:
still these are constant which already exist in the package you use.

Reply:
I agree that whatever can be imported is better than copy-paste

Reply:
They are internal names in pyds8k, I don't expect them to be used in other projects.

Reply:
If you don't expect them to be used in other projects, why did you copy (all of) them here?
And if you need (some of) them here, why don't you expect them to be used in other projects?

Reply:
Let's distinguish different const situations:
1. for consts that you are not using now (e.g. `SYSTEM_CAPACITY`), remove them.
2. for consts that are completely the same in pyds8k (e.g. `SYSTEM_ID`), please [import](https://github.com/IBM/pyds8k/blob/ade6e0e5a3f712e5b6c2d78769824bf8afcc0057/pyds8k/resources/ds8k/v1/common/attr_names.py) them.
3. for consts that does not exist in pyds8k (e.g. `INVALID_CREDENTIALS`?), if you cannot add them to pyds8k AND need to use them now, leave them here for now.

Reply:
removed the unused consts, but I don't want to import them from pyds8k as I explained above.

Reply:
regarding your above explanation:
I understand that you want an abstraction between storage and pyds8k fields, but this abstraction is the responsibility of pyds8k.
You wrote: "the ones here stand for the fields from pyds8k" - And I say: "the ones that stand for the fields from pyds8k, should be defined there, not here".
pyds8k should expose an API, API should consist of both functions and consts to be used with those functions.
If pyds8k does not provide separate consts to be imported, then there is no abstraction, so CSI has to import the storage side fields, until pyds8k will decide to create an abstraction.

Reply:
I don't think pyds8k should defined the consts.

Reply:
Then how can the CSI client "know" them? they are relevant only to the pyds8k API.
It doesn't make sense for every pyds8k consumer to re-define them, pyds8k is the one that requires them, so it needs to expose them.
If you don't agree with me, ask for other reviewers opinion, so we can settle this.

Reply:
These consts are just json response keys, I can't find any reason to define all of them on pyds8k side. Take pyxcli as an example, if get_volume returns {"name": a, "size", 1}, do you need the define "name", "size"  on pyxcli as VOLUME_NAME VOLUME_SIZE and import them from pyxcli?

Reply:
`get_volume` does not return a `dict`, it returns an [object](https://github.com/IBM/ibm-block-csi-driver/blob/e91e9edf489ae2a4a942d1c9c2324d8dd2290617/controller/array_action/array_mediator_xiv.py#L78) with members, so no need for consts with pyxcli.
if it's a json, can't you deserialize it to an API class?

Reply:
I think you misunderstand my words, get_volume is just an example, I mean the api in pyxcli to get a volume (In fact, it is called vol_list(vol=vol_name)), and yes, it return a dict (a munch actually, but they are similar).
it returns a dict, not a json, correct it.

Reply:
I see.
Here are the following options, from best to worst:
1. return an object of a structure-defined class.
2. return a Munch
3. return a dict (that requires many hard-coded consts, which their definition location is controversial)

The code is currently at option 3. if it cannot be 1, then at least make it 2, to avoid the hard-coded definitions location issue.

Reply:
it is indeed strange that an error dict is returned instead of an exception being raised, but the `RESTClient` can validate and raise the appropriate exception in case of ("silent") non-valid resource, so at least the mediator will get a consistent type and specific exceptions.

Reply:
Maybe i didn't make it clear. Usually exception will be handled properly. But ds8k REST APIs support bulk creation, and for partially successful request, pyds8k will return a list like `[vol1_obj, vol2_obj, vol3_error_dict, ...]`

Reply:
Since the mediator doesn't need bulk creation, it should use a consistent API (i.e. get only one type of object, or except an exception)

Reply:
Yes, as I said, it is only for partially successful requests, currently there is no such requests, so it will raise errors.

Reply:
so to your question, yes, please change the usage to be of the form: `vol.name` (instead of `vol[NAME]`)

Reply:
ok, accepted. @oriyarde , please check it again.

Reply:
LGTM